tests/post_training/pipelines/lm_weight_compression.py

```diff
@@ -192,6 +178,16 @@ def _compress(self):
         """
         Actual call of weight compression
         """
+        if self.backend == BackendType.TORCH:
+            """If Backend is TORCH (Assuming that it's INT8 compression), don't use a dataset as it's Unsupported"""
+            self.compressed_model = nncf.compress_weights(
+                self.model,
```

Comment:
It looks strange that dataset is None, because this is required parameter for PyTorch model. Please check that you feed PyTorch model here.

Reply:
Yeah agreed. But when I ran the command you provided it threw an error saying that "INT8 doesn't support Dataset option, set it to `None`

Reply:
FYI: https://github.com/openvinotoolkit/nncf/blob/fa1a4ceabee06e10f7dfb42ce12bdd9b04ed4b79/nncf/quantization/quantize_model.py#L382

Reply:
> FYI:
> 
> https://github.com/openvinotoolkit/nncf/blob/fa1a4ceabee06e10f7dfb42ce12bdd9b04ed4b79/nncf/quantization/quantize_model.py#L382

Interesting. Let me try to wrap the model as done in the referenced file, and then run it with a passed dataset.
Also, here's the output with a passed dataset currently: 
```bash
FAILED tests/post_training/test_quantize_conformance.py::test_weight_compression[tinyllama_int8_data_free_backend_TORCH] - Failed: INT8 modes do not support `all_layers`, `sensitivity_metric`, `awq` and `dataset
` options. Set them to None.
```

Reply:
@alexsu52 
They also [set the dataset](https://github.com/openvinotoolkit/nncf/blob/fa1a4ceabee06e10f7dfb42ce12bdd9b04ed4b79/nncf/quantization/quantize_model.py#L408C1-L409C27) to `None` after wrapping the model

I think while `dataset` is necessary for `PyTorch`, it has to be set to `None` for `INT8` compression, I will do just that in the code and update you

Reply:
@alexsu52 
Wrapping the model as follows, doesn't work (I tried various solutions, such as adding checks for compression type, checking for `Model` object and etc.)

### Are we **sure** that setting `Dataset` to `None` is incorrect? 

# Code
```python
    def _compress(self):
        """
        Actual call of weight compression
        """
        if self.backend == BackendType.TORCH:
            from nncf.torch.model_creation import is_wrapped_model
            from nncf.torch.model_creation import wrap_model

            if not is_wrapped_model(self.model):
                example_input = next(iter(self.calibration_dataset.get_inference_data()))
                self.model = wrap_model(self.model, example_input=example_input, trace_parameters=True)
                self.calibration_dataset = None

        self.compressed_model = nncf.compress_weights(
            self.model,
            dataset=self.calibration_dataset,
            **self.compression_params,
        )
```

# Error
```bash
FAILED tests/post_training/test_quantize_conformance.py::test_weight_compression[tinyllama_int8_data_free_backend_TORCH] - Failed: 'Model' object has no attribute 'named_modules'
```

Reply:
Your statement is very interesting. Let's try to figure out.
At the first I would recommend to check that your environment is correct to run the code from the issue:
```
import nncf
import transformers

MODEL_ID = "tinyllama/tinyllama-1.1b-step-50k-105b"

tokenizer = transformers.AutoTokenizer.from_pretrained(MODEL_ID)
model = transformers.AutoModelForCausalLM.from_pretrained(MODEL_ID, torch_dtype=torch.float32, device_map="cpu")

text = "The TinyLlama project aims to pretrain a 1.1B Llama model on 3 trillion tokens."
token = tokenizer(text, max_length=500, return_tensors="pt", truncation=True)
inputs = {"input_ids": token["input_ids"], "attention_mask": token["attention_mask"]}

compressed_model = nncf.compress_weights(model, dataset=nncf.Dataset([inputs]))
``` 

I should get the following log:
![image](https://github.com/openvinotoolkit/nncf/assets/3229971/10f78e10-057d-4314-9db7-f353aa7e92b2)

If you reproduced the result of the provided code in your environment then the environment is correct.
Next step to provide a short code script to reproduce your issue, then I will try to reproduce on my side.

Reply:
@alexsu52 Funnily enough, I just did the exact thing you asked for xD

# Result
---
![image](https://github.com/openvinotoolkit/nncf/assets/80326762/70485e86-f74b-4c5f-99a0-287648b6dd55)

Reply:
@alexsu52 do you require anything else to reproduce the problem?

The code is in the above comment

Reply:
@AdiKsOnDev please, provide standalone reproducer to just copy and run as I provided to you above.

Reply:
Just confirming, you want me to send the code that I ran to check the environment, right? The same code you provided?

Reply:
@alexsu52 Sorry for taking a while, I am not that familiar with the pipeline yet

>[!WARNING]
> This code is pretty stupid and **might** fail even after fixing the error below.
> This code is provided purely for the sake of debugging the error shown below:

```bash
Weight compression...
'Model' object has no attribute 'named_modules'
```



```python
import pathlib
import time

from nncf.parameters import CompressWeightsMode
from tests.post_training.pipelines.base import BackendType
from tests.post_training.pipelines.lm_weight_compression import LMWeightCompression

pipeline = None
err_msg = None
test_model_param = None
start_time = time.perf_counter()
output_dir = pathlib.Path("./model/")
data_dir = pathlib.Path("./tests/post_training/data/")

try:
    pipeline_cls = LMWeightCompression
    pipeline_kwargs = {
        "reported_name": "tinyllama_int8_data_free",
        "model_id": "tinyllama/tinyllama-1.1b-step-50k-105b",
        "backend": BackendType.TORCH,
        "compression_params": {
            "mode": CompressWeightsMode.INT8_ASYM,
            "all_layers": None,
            "awq": None,
            "sensitivity_metric": None,
        },
        "output_dir": output_dir,
        "data_dir": data_dir,
        "reference_data": {"median": 0.9, "num_int4": 228, "num_int8": 84},
        "no_eval": True,
        "run_benchmark_app": False,
    }
    pipeline: LMWeightCompression = pipeline_cls(**pipeline_kwargs)
    pipeline.prepare()
    pipeline.compress()
except Exception as e:
    print(e)

```

# The problematic code in `LMWeightCompression` (I will commit it in a sec)
```python
    def _compress(self):
        """
        Actual call of weight compression
        """
        if self.backend == BackendType.TORCH:
            from nncf.torch.model_creation import is_wrapped_model
            from nncf.torch.model_creation import wrap_model

            if not is_wrapped_model(self.model):
                example_input = next(iter(self.calibration_dataset.get_inference_data()))
                self.model = wrap_model(self.model, example_input=example_input, trace_parameters=True)
                self.calibration_dataset = None

        self.compressed_model = nncf.compress_weights(
            self.model,
            dataset=self.calibration_dataset,
            **self.compression_params,
        )
```

Reply:
@alexsu52 Hi, any updates?

Reply:
I run your code with some modification and got the following trace stack:
```
tests/post_training/pipelines/base.py:311: in run
    self.compress()
tests/post_training/pipelines/lm_weight_compression.py:150: in compress
    self.run_info.compression_memory_usage = memory_usage(self._compress, max_usage=True)
../../venv/nncf_openvino/lib/python3.10/site-packages/memory_profiler.py:379: in memory_usage
    returned = f(*args, **kw)
tests/post_training/pipelines/lm_weight_compression.py:201: in _compress
    self.model = wrap_model(self.model, example_input=example_input, trace_parameters=True)
nncf/torch/model_creation.py:332: in wrap_model
    with training_mode_switcher(model, is_training=False):
/usr/local/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
nncf/torch/utils.py:257: in training_mode_switcher
    saved_state = save_module_state(model)
nncf/torch/utils.py:236: in save_module_state
    return _ModuleState(module)
```
`wrap_model` crashed because you feed OpenVINO model instead of PyTorch model.
![image](https://github.com/openvinotoolkit/nncf/assets/3229971/a8d86c42-6a4f-4d1e-9d88-a789ea862150)

i added additional check in `wrap_model` function: https://github.com/openvinotoolkit/nncf/pull/2651

Reply:
@alexsu52 Hey, I am trying to do the following:
```python
class LMWeightCompression(BaseTestPipeline):
    """Pipeline for casual language models from Hugging Face repository"""

    MODEL_NAME = "openvino_model.xml"
    MODEL_FUNC = OVModelForCausalLM

    def prepare_model(self) -> None:
        is_stateful = self.params.get("is_stateful", False)
        if self.backend == BackendType.TORCH:
            self.MODEL_NAME = "torch_model.xml"
            self.MODEL_FUNC = AutoModelForCausalLM
            MODEL_SPECIFIC_PARAMS = {}
        else:
            MODEL_SPECIFIC_PARAMS = {"export": True, "compile": False, "stateful": is_stateful}
        if is_stateful:
            self.fp32_model_dir = self.fp32_model_dir.parent / (self.fp32_model_dir.name + "_sf")
        if not (self.fp32_model_dir / self.MODEL_NAME).exists():
            # export by model_id
            self.model_hf = self.MODEL_FUNC.from_pretrained(
                self.model_id, load_in_8bit=False, **MODEL_SPECIFIC_PARAMS
            )
            self._dump_model_fp32()
        else:
            # no export, load from IR. Applicable for sequential run of test cases in local environment.
            self.model_hf = self.MODEL_FUNC.from_pretrained(
                self.fp32_model_dir, trust_remote_code=True, load_in_8bit=False, **MODEL_SPECIFIC_PARAMS
            )
        self.model = self.model_hf.model
```
And it works so far, but I need some substitution for `_save_config()` method (See the error message below)

```bash
FAILED tests/post_training/test_quantize_conformance.py::test_weight_compression[tinyllama_int8_data_free_backend_TORCH] - Failed: 'LlamaForCausalLM' object has no attribute '_save_config'
```

Do you know any methods that have the same functionality? Or am I going in a completely wrong direction?

Reply:
Hi @alexsu52, a quick question; does `.compress()` **always** expect an `OpenVINO` model? After wrapping the `torch` model and starting the compression, this is the error I get:
```bash
Failed: serialize(): incompatible function arguments. The following argument types are supported:
    1. (model: ov::Model, xml_path: object, bin_path: object = '', version: str = 'UNSPECIFIED') -> None
```
I also get the following warning:
```bash
tests/post_training/test_quantize_conformance.py huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
```

Reply:
Hi! 

I have reviewed your PR and It looks like you have updated almost all parts of the pipeline and the implementation now tests the compression of the PyTorch model.

I guess, the error relatives to that `ov.serialize` does not support PyTorch model. You should convert compressed PyTorch to OpenVINO using `ov.convert_model` and then call `ov.serialize` for the converted model.

Reply:
Okay great, I was starting to think that I'm going insane😂
Thank you, will send a commit in a bit

Reply:
@alexsu52 the thing is, I was thinking about doing this yesterday and I started wondering; what's the point in calling `AutoModelForCausalLM` instead of ` `OVModelForCausalLM` if we end up converting it back to ov model either way?

Reply:
@alexsu52 Do you know a fix for this, by chance? Passing the attention mask didn't help, adding attn_implementation also didn't help. (I also tried changing example_input a number of times, this is the closest I could get to a proper result)
## Code
```python
        self.compressed_model = ov.convert_model(
            self.compressed_model, example_input=torch.rand(1, 3, 224, 224).to(torch.long)
        )
```

## Error
```bash
E           Failed: Couldn't get TorchScript module by tracing. With exception:
E           Attention using SDPA can not be traced with torch.jit.trace when no attention_mask is provided. To solve this issue, please either load your model with the argument `attn_implementation="eager"` or p
ass an attention_mask input when tracing the model.

```

Reply:
@alexsu52 This is the output after adding "attn_implementation=eager" to the model:
```bash
E :Dictionary inputs to traced functions must have consistent type. 
Found Tensor and Tuple[Tuple[Tensor, Tensor], 
Tuple[Tensor, Tensor], Tuple[Tensor, Tensor], Tuple[Tensor, Tensor], Tuple[Tensor, Tens...
```

Reply:
Good morning @alexsu52, just want to update you in case you happen to know a solution;
So I spent the whole evening and morning trying some fixes I could come across. 
# Things I Tried:
- I tried converting the model to `ScriptModule` via `torch.jit.script()` before converting to `OpenVINO`
- Tried passing a random `torch.Tensor` as an example_input
- Tried passing the tokenized example input 
- Passed the `nncf.Dataset`
- Added `attn_implementation="eager"` to the model
- Tried converting the model **BEFORE** compression
- Tried omitting the `ov.serialize()` for TORCH models
- And tried converting to `ONNX`, before converting it to `OpenVINO`
- and some other solutions that are similar to the previous ones
> [!NOTE]
> I think it's mainly the example_input that's wrong. As long as I can find a way to (Maybe) reshape the input or 
> something like that, I'll be done with the PR.

Reply:
@alexsu52 I do not think it is possible to run `torch.jit.trace()` on `CausalModels`'s. Therefore, it's not possible to convert them using `ov.convert_model()`. I'll try to look for other workaround for the `ov.serialize()` problem.